pallets/inflation/src/lib.rs

```diff
@@ -0,0 +1,604 @@
+            let dapp_reward_pool_per_era = dapps_emission / build_and_earn_eras_per_cycle;
+
+            // 3.3. Staking reward pools per era
+            let base_staker_reward_pool_per_era =
+                base_stakers_emission / build_and_earn_eras_per_cycle;
+            let adjustable_staker_reward_pool_per_era =
+                adjustable_stakers_emission / build_and_earn_eras_per_cycle;
+
+            // 3.4. Bonus reward pool per period
+            let bonus_reward_pool_per_period = bonus_emission / periods_per_cycle;
+
+            // 4. Block at which the inflation must be recalculated.
+            let recalculation_block = now.saturating_add(T::CycleConfiguration::blocks_per_cycle());
+
+            // 5. Return calculated values
+            (
+                max_emission,
+                InflationConfiguration {
+                    recalculation_block,
+                    collator_reward_per_block,
+                    treasury_reward_per_block,
+                    dapp_reward_pool_per_era,
+                    base_staker_reward_pool_per_era,
+                    adjustable_staker_reward_pool_per_era,
+                    bonus_reward_pool_per_period,
+                    ideal_staking_rate: params.ideal_staking_rate,
+                },
+            )
+        }
+    }
+
+    impl<T: Config> StakingRewardHandler<T::AccountId> for Pallet<T> {
+        fn staker_and_dapp_reward_pools(total_value_staked: Balance) -> (Balance, Balance) {
+            let config = ActiveInflationConfig::<T>::get();
+            let total_issuance = T::Currency::total_issuance();
+
+            // First calculate the adjustable part of the staker reward pool, according to formula:
+            // adjustable_part = max_adjustable_part * min(1, total_staked_percent / ideal_staked_percent)
+            // (These operations are overflow & zero-division safe)
+            let staked_ratio = Perquintill::from_rational(total_value_staked, total_issuance);
+            let adjustment_factor = staked_ratio / config.ideal_staking_rate;
+
+            let adjustable_part = adjustment_factor * config.adjustable_staker_reward_pool_per_era;
+            let staker_reward_pool = config
+                .base_staker_reward_pool_per_era
+                .saturating_add(adjustable_part);
+
+            (staker_reward_pool, config.dapp_reward_pool_per_era)
+        }
+
+        fn bonus_reward_pool() -> Balance {
+            ActiveInflationConfig::<T>::get().bonus_reward_pool_per_period
+        }
+
+        fn payout_reward(account: &T::AccountId, reward: Balance) -> Result<(), ()> {
+            let mut tracker = SafetyInflationTracker::<T>::get();
+
+            // This is a safety measure to prevent excessive minting.
+            // TODO: discuss this in review with the team. Is it strict enough? Should we use a different approach?
+            tracker.issued.saturating_accrue(reward);
+            ensure!(tracker.issued <= tracker.cap, ());
```

Comment:
It's still possible for a single account (or multiple ones) to claim excessive amounts of rewards, if a bug happens in dApp staking. But it's not indefinite.

> nitpick, we can have a method (called maybe issue(amount)) on tracker struct itself for increasing the issued with this check there itself.

We could add that, but we cannot reuse it in `on_timestamp` since that has to be infallible 🙂